Question 1, part 1

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import math
from sympy import solve

trap1=0
trap2=0
trapf=0

def CTR(N,a,b,f,trap1,trap2,trapf):
    h = float((b-a) / N)
    for i in range(1, int(N)):
        trap1 += (f(a + i*h))
        trap2 = trap1 + (0.5*f(a) + 0.5*f(b))
        trapf = float(h*trap2)
    return(trapf);

def func(x): return 1/((1+x)**2)
test1 = CTR(20,0,2,func,trap1,trap2,trapf)
test2 = CTR(40,0,2,func,trap1,trap2,trapf)
test3 = CTR(80,0,2,func,trap1,trap2,trapf)
print("test1 = ",test1)
print("test1 = ",test2)
print("test1 = ",test3)

def integral(x): return -1/(1+x)
I_f = integral(2) - integral(0)
error1 = np.abs(I_f - test1)
error2 = np.abs(I_f - test2)
error3 = np.abs(I_f - test3)
print("error1 = ", error1)
print("error2 = ", error2)
print("error3 = ", error3)

print("error1 / error2 = ", error1 / error2, " which is approximately 4 = 2^2. Therefore rate of convergence is 2.")
print("error2 / error3 = ", error2 / error3, " which is approximately 4 = 2^2. Therefore rate of convergence is 2.")


test1 =  0.6682683087950135
test1 =  0.6670676941291325
test1 =  0.6667669623471976
error1 =  0.0016016421283467919
error2 =  0.0004010274624657306
error3 =  0.00010029568053082638
error1 / error2 =  3.993846502429141  which is approximately 4 = 2^2. Therefore rate of convergence is 2.
error2 / error3 =  3.9984519806161822  which is approximately 4 = 2^2. Therefore rate of convergence is 2.


Question 1, part 2

In [2]:
def func2(x): return np.sqrt(x)
test16 = CTR(16,0,1,func2,trap1,trap2,trapf)
test32 = CTR(32,0,1,func2,trap1,trap2,trapf)
test64 = CTR(64,0,1,func2,trap1,trap2,trapf)
test128 = CTR(128,0,1,func2,trap1,trap2,trapf)
print("when N = 16, T = ",test16)
print("when N = 32, T = ",test32)
print("when N = 64, T = ",test64)
print("when N = 128, T = ",test128)

def integ(x): return 2*(x**(3/2)) / 3
I_f2 = integ(1) - integ(0)
error16 = np.abs(I_f2 - test16)
error32 = np.abs(I_f2 - test32)
error64 = np.abs(I_f2 - test64)
error128 = np.abs(I_f2 - test128)
print("error16 = ",error16)
print("error32 = ",error32)
print("error64 = ",error64)
print("error128 = ",error128)

print("error16 / error32 = ", error16 / error32,)
print("error32 / error64 = ", error32 / error64)
print("error64 / error128 = ", error64 / error128)


when N = 16, T =  0.6635811968772283
when N = 32, T =  0.6655589362789418
when N = 64, T =  0.6662708113785066
when N = 128, T =  0.6665256572968262
error16 =  0.0030854697894383554
error32 =  0.0011077303877248257
error64 =  0.0003958552881599964
error128 =  0.0001410093698404058
error16 / error32 =  2.7853978040411267
error32 / error64 =  2.7983215605726715
error64 / error128 =  2.807297760482334


There isn't a second order convergence to the exact value of the integral because the second derivative of $f(x)$ diverges at $0$, so the errors decay slowly for $x$ values close to $0$.

Question 2, part 1

In [3]:
def func3(x): return np.cos(x**2)

Th1 = lambda N: CTR(N, 0, np.sqrt(np.pi/2), func3,trap1,trap2,trapf)
Th2 = lambda N: CTR(2*N, 0, np.sqrt(np.pi/2), func3,trap1,trap2,trapf)
Th4 = lambda N: CTR(4*N, 0, np.sqrt(np.pi/2), func3,trap1,trap2,trapf)

Fc1 = lambda N: (Th2(N) - Th1(N)) / (Th4(N) - Th2(N))
for N in range(1, 1000):
       if (3.999999 <= Fc1(N) <= 4.000001):
        print("N=",N)

N= 736
N= 751
N= 754
N= 757
N= 763
N= 765
N= 767
N= 769
N= 772
N= 777
N= 779
N= 780
N= 782
N= 783
N= 784
N= 786
N= 787
N= 788
N= 791
N= 792
N= 793
N= 794
N= 795
N= 796
N= 797
N= 802
N= 803
N= 804
N= 805
N= 810
N= 811
N= 814
N= 815
N= 816
N= 817
N= 818
N= 819
N= 820
N= 821
N= 822
N= 823
N= 824
N= 825
N= 826
N= 827
N= 828
N= 829
N= 830
N= 831
N= 833
N= 834
N= 835
N= 836
N= 837
N= 838
N= 839
N= 841
N= 842
N= 843
N= 845
N= 847
N= 848
N= 849
N= 850
N= 851
N= 852
N= 853
N= 854
N= 855
N= 856
N= 857
N= 858
N= 859
N= 860
N= 862
N= 863
N= 864
N= 865
N= 866
N= 867
N= 868
N= 870
N= 871
N= 872
N= 873
N= 874
N= 875
N= 876
N= 877
N= 878
N= 879
N= 880
N= 881
N= 882
N= 883
N= 884
N= 885
N= 886
N= 888
N= 889
N= 890
N= 891
N= 892
N= 893
N= 894
N= 895
N= 897
N= 898
N= 899
N= 900
N= 901
N= 902
N= 903
N= 904
N= 905
N= 906
N= 907
N= 908
N= 909
N= 910
N= 911
N= 912
N= 913
N= 914
N= 915
N= 916
N= 917
N= 918
N= 919
N= 920
N= 921
N= 922
N= 923
N= 924
N= 925
N= 926
N= 927
N= 928
N= 929
N= 930
N= 931
N= 932
N= 933

In [4]:
#just to check
print("When N = 736, q(h) = ", Fc1(736), " which is approximately 4.") 

When N = 736, q(h) =  4.000000968745322  which is approximately 4.


Let's pick $N = 736$. Then a value of $h$ is  $h = \frac{b-a}{N}\ = \frac{\sqrt\frac{\pi}{2}}{736} = 0.001702872469 $.

Question 2, part 2

In [13]:
Int = 0.977451424291
T_h = CTR(736,0,np.sqrt(np.pi/2),func3,trap1,trap2,trapf)
print("I - Th = ", np.abs(Int - T_h))

I - Th =  6.057212814658541e-07


Question 2, part 3

$I - T_h = \frac{4}{3}(T_\frac{h}{2} - T_h)$ so we can use our error approximation to solve for $S_h$.

In [25]:
Sh = T_h + (Int - T_h)
print("Sh = ", Sh)


Sh =  0.977451424291


Question 2, part 4

Simpson's rule is more accurate than the trapezoidal rule and converges faster to the exact value because it uses quadratic approximations rather than linear approximations. It divides the intervals further than the trapezoidal rule in order to give a more accurate result

Question 2, part 5

In [12]:
Th8 = lambda N: CTR(8*N, 0, np.sqrt(np.pi/2), func3,trap1,trap2,trapf)

Sh1 = lambda N: Th1(N) + (4/3)*(Th2(N) - Th1(N))
Sh2 = lambda N: Th2(N) + (4/3)*(Th4(N) - Th2(N))
Sh4 = lambda N: Th4(N) + (4/3)*(Th8(N) - Th4(N))

for N in range (1,300):
    if (15.999999 <= (((Sh2(N) - Sh1(N))) / (Sh4(N) - Sh2(N)) ) <= 16.000001):
        print("N = ", N)

N =  227
N =  234


Let $N = 234$ since $N$ must be even for Simpson's rule. Then $h = \frac{b-a}{N}\ = \frac{\sqrt\frac{\pi}{2}}{234} = 0.005356043322$.